### PR TITLE
Abiv2

### PR DESCRIPTION
**Pull request by vkgnosis**
Edit: Ready for review, scroll down for update.

Nick pointed out that there is an ethabi fork that already supports abiv2. This PR is using this.

Where I am currently stuck is on the tokenization for web3. Web3 has several related traits https://docs.rs/web3/0.15.0/web3/contract/tokens/index.html . The trait that is usually accepted and used by ethcontract is `Tokenize`.
In my solidity example we have
```solidity
struct S {
  uint8 u0;
  uint16 u1;
}

function takeStruct(S calldata s) public view {}
```
In the abi this becomes `Tuple(uint8, uint16)` which gets turned by ethcontract in this PR into `s : (u8, u16)`.

Web3 has one trait `Tokenizable` for things that can get turned into a single `Token` and another trait `Tokenize` that turns things into `Vec<Token>`. The latter is implemented probably for convenience  so that users of web3 can pass tuples of different tokenziable things to it. This has now become problematic because it prevents tuples of Tokenizable to themself implement Tokenizable (by turning into Token::Tuple) because of ambiguous implementations. Imo the correct way to fix this is to get rid of the Tokenize trait so that we can have the mentioned tuple impl. This involves getting this merged in web3 first.
I am not sure if there is a way to achieve the same thing without touching web3. The only alternative I see is to duplicate Tokenize into ethcontract and implemented it correctly for tuples and then pass our Tokenize around instead of web3's.

### CODE REVIEW

**Comment by fleupold** [comment]
I don't fully understand the issue (I checked out the code but couldn't find concrete errors).

[Tokenizable](https://docs.rs/web3/0.15.0/web3/contract/tokens/trait.Tokenizable.html#impl-Tokenizable-for-%5BT%3B%202%5D) requires two methods: `from_token` and `into_token`.

[Tokenize](https://docs.rs/web3/0.15.0/web3/contract/tokens/trait.Tokenize.html) requires one method: `into_tokens`. Where is the ambiguous implementation?

What I don't understand is how you would be able to implement `Tokenizable` on the raw Tuple as neither the type not the trait are local to your crate? Wouldn't we have to add this implementation to web3?

**Comment by vkgnosis** [comment]
Probably not worth your time to investigate too deeply, Felix. I was mostly thinking out loud in case @nlordell had encountered this before. Will try to explain it again:
Web3 has a trait Tokenizable that is implemented for basic types like u8, String, etc. The trait is also implemented for slices of basic types because those can map to Token::Array. In the same vein you would expect the trait to be implemented for tuples of itself because those map to Token::Tuple.
But actually web3 has a different trait Tokenize that is implemented on tuples of Tokenizable. The previous trait is able to turn T into a single Token and this trait is able to turn T into Vec<Token>. I think the reason for having this is to make it easy to pass tokens to representing multiple function arguments. The problem is that with this setup if tuples of Tokenizable were themself Tokenizable it would be ambiguous whether they would encode into a single Token::Tuple or a Vec<Token>. This is why tuples of Tokenizable aren't and cannot be Tokenizable. Which means with web3's api you cannot pass something like ((u8; u16); u32) because (u8; u16) is not Tokenizable so the whole thing is not Tokenizable.
As you say fixing this properly requires a change in web3. Another solution I see now is that we can make our own Tokenizable trait that doesn't have these issues. That might be useful as an immediate fix while discussion on what is best to do in web3 continues in form of a PR there.

**Comment by nlordell** [comment]
> I was mostly thinking out loud in case @nlordell had encountered this before.

This rings a bell. I think this is why I abandoned implementing this a while back.

I think the issue is that the `web3` traits are insufficient for modelling nested `Tokenizable` doesn't support nesting as you described.

**Comment by vkgnosis** [comment]
I am working on a PR for web3 that would fix this but it has to introduce some breakage there. Until that is merged (if ever) we should be able to duplicate the same trait changes into ethcontractract because we manually call ethabi::encode / decode anyway without going through web3 for that, so we don't have to use the same trait to create Tokens as web3 does.

**Comment by vkgnosis** [comment]
I've moved the tokenizaton related traits from rust-web3 into ethcontract. This makes them simpler (than the rust-web3 version) because ethcontract uses them for fewer purposes than rust-web3 and gives us more control over them. I feel this is what we should do mid-term until we see the route web3 and ethabi take.

Still WIP because I need to add error handling, documentation and more tests for abi v2.

File: Cargo.toml
Comment (nlordell) [inline Cargo.toml]: I think patches are only respected at a top level right? So this patch would also have to be applied for `gp-v2-contracts`

File: ethcontract-derive/src/lib.rs
Comment (vkgnosis) [inline ethcontract-derive/src/lib.rs:345]: Needed because we have to create an instance of the struct which involves touching the deprecated `constant` field.

File: ethcontract/src/int.rs
Comment (vkgnosis) [inline ethcontract/src/int.rs:1218]: Moved to `tokens.rs`

File: ethcontract/src/log.rs
Comment (vkgnosis) [inline ethcontract/src/log.rs:161]: Needed because the type would come from a different version of ethabi. This is just a copy of web3's implementation of `topic_filter`.

File: examples/truffle/contracts/AbiTypes.sol
Comment (vkgnosis) [inline examples/truffle/contracts/AbiTypes.sol:90]: See the commit message for an explanation of why this is needed.

**Comment by vkgnosis** [comment]
This is ready for review now. PR is big but I don't see a way to split this into semantically useful parts. 80% of the changes are the new file `tokens.rs` which itself mostly comes from rust-web3. I recommend reading the commit message too for some context.

You can see the change I had to make to ethabi here https://github.com/vkgnosis/ethabi/commits/ethcontract .

File: ethcontract-derive/src/lib.rs
Comment (fleupold) [inline ethcontract-derive/src/lib.rs:354]: Does this come from the ethABI upgrade? If so maybe this could have been separated out into its own PR to make the ethABI change more local.

File: ethcontract-common/Cargo.toml
Comment (vkgnosis) [inline ethcontract-common/Cargo.toml]: Not sure what the best way to distribute this is. Could keep on git, push to crates, include the whole crate in this repo (maybe as a submodule).

Comment (nlordell) [reply ethcontract-common/Cargo.toml]: > Could keep on git

The issue there is that we can't publish to Crates.io then.

File: ethcontract-derive/src/lib.rs
Comment (nlordell) [reply ethcontract-derive/src/lib.rs:345]: Can the allow be moved directly over `constant: false`?

File: ethcontract/src/contract/method.rs
Comment (fleupold) [inline ethcontract/src/contract/method.rs]: naming nit: Maybe Returnable is more similar to Detokenizable? I like the clarity that this is about a Return Type.

File: ethcontract/src/tokens.rs
Comment (fleupold) [inline ethcontract/src/tokens.rs]: nit: this is also used in Return

File: ethcontract-generate/src/contract/types.rs
Comment (fleupold) [inline ethcontract-generate/src/contract/types.rs:47]: As a future feature it would be super nice if this could autogenerate a named struct for the type. Do we have access to this type of info in the artefact?

File: ethcontract/src/tokens.rs
Comment (fleupold) [inline ethcontract/src/tokens.rs:69]: Isn't this missing the FixedBytes arm?

File: ethcontract/src/tokens.rs
Comment (fleupold) [inline ethcontract/src/tokens.rs]: Do I understand correctly, that the problem with the Web3 Tokenize trade is this line:

> impl<T: Tokenizable> Tokenize for T

Since we don't implement this we don't get into the ambiguity that you mentioned? 

If that's the case maybe it would be better to try and leave the naming mostly similar and just remove this one implementation and add the `impl_single_tokenize_for_tuple` lines. This way the change is slightly easier to compare and might have a chance of being upstream compatible if we can get it merged upstream.

File: examples/truffle/contracts/AbiTypes.sol
Comment (fleupold) [reply examples/truffle/contracts/AbiTypes.sol:90]: Could you point out which stricter size verification is causing this error to happen now?

**Review by fleupold — COMMENTED** [review]
Just some comments, as I'm still not sure I fully understand what's going on. Leaving the thorough review up to @nlordell

File: ethcontract/Cargo.toml
Comment (nlordell) [inline ethcontract/Cargo.toml:28]: I don't see this being used anywhere? Maybe I missed something.

File: ethcontract/src/contract/method.rs
Comment (nlordell) [inline ethcontract/src/contract/method.rs]: Does this mean we don't support `MultiTokenize` return types? ABI v2 allows returning structs for example.

**Review by nlordell — COMMENTED** [review]
In general, the code looks good.

I still haven't wrapped my head around the `SingleTokenize` and `MultiTokenize` traits. I'm going to take a second look with fresh eyes on Monday.

File: ethcontract/src/contract/method.rs
Comment (nlordell) [reply ethcontract/src/contract/method.rs]: I see `SingleTokenize` supports `Tuple` as well.

File: ethcontract-derive/src/lib.rs
Comment (vkgnosis) [reply ethcontract-derive/src/lib.rs:345]: Tried that but doesn't work.

File: examples/truffle/contracts/AbiTypes.sol
Comment (vkgnosis) [reply examples/truffle/contracts/AbiTypes.sol:90]: https://github.com/tomusdrw/rust-web3/blob/e6d044a28458be9a3ee31108475d787e0440ce8b/src/contract/tokens.rs#L234

In the web3 version it does not check the value of the integer versus the min/max of the type.

File: ethcontract-derive/src/lib.rs
Comment (vkgnosis) [reply ethcontract-derive/src/lib.rs:354]: It does. But can't separate it out because then the ethabi Token types come from different versions so it wouldn't compile.

File: ethcontract-generate/src/contract/types.rs
Comment (vkgnosis) [reply ethcontract-generate/src/contract/types.rs:47]: Agree it would be nice but not possible with the information we get from ethabi. According to the offical description of the json abi artifact this isn't possible because these structs-as-tuples are indistinguishable from plain tuples. In practice there is an undocumented field in the json that might hold enough information.

File: ethcontract/src/tokens.rs
Comment (vkgnosis) [reply ethcontract/src/tokens.rs:69]: This is intentional because we always translate Token::Bytes to Vec<u8> so it should never be the case that we call this with FixedBytes. This is an example where we can be stricter than the rust-web3 version which presumably wants to give users more flexibility by allowing either version of bytes.

File: ethcontract/src/tokens.rs
Comment (vkgnosis) [reply ethcontract/src/tokens.rs]: Imo the web3 version and our version have different goals so I am unsure about whether we will go back to the web3 version. rust-web3 does not support a type safe api so it gives users more flexibility in the way they can provide rust types. We have full control over the rust interface so we can do some things differently and don't have to provide ergonomics.

Web3 *should* also add some kind of fix for tuples and I have made a PR for that but it is unclear whether that implementation would also work for ethcontract. That is why I treated this as more of  a from scratch kind of thing and renamed traits when I felt the old names were bad.

https://github.com/tomusdrw/rust-web3/pull/464 has another attempt at explaining the problem.

File: ethcontract/Cargo.toml
Comment (vkgnosis) [reply ethcontract/Cargo.toml:28]: It's used in `tokens.rs`.

File: ethcontract/src/contract/method.rs
Comment (vkgnosis) [reply ethcontract/src/contract/method.rs]: The difference is how this presented in the abi json. In ethabi the decode and encode functions work on Vec<Token>. This corresponds to function arguments being a json list and not a top level tuple, so they should encode into and from Vec<Token>. Return types are technically a list too in the json but as far as I can tell they should always be a tuple so a single Token.

Comment (vkgnosis) [reply ethcontract/src/contract/method.rs]: I wonder if Vec<Token> and Tuple encode to the same bytes. In that case we might be able to get rid of MultiTokenize by pretending that our arguments form a single tuple all together.

Comment (nlordell) [reply ethcontract/src/contract/method.rs]: > I wonder if Vec and Tuple encode to the same bytes

I think `Vec` encodes to `offest -> length || item0 || item1 || ...` while `Tuple` encodes to `offset -> item0 || item1 || ...`.

File: ethcontract/Cargo.toml
Comment (nlordell) [reply ethcontract/Cargo.toml:28]: The `token.rs` wasn't expanded when I did Ctrl-F :stuck_out_tongue:

File: ethcontract-generate/src/contract/types.rs
Comment (nlordell) [reply ethcontract-generate/src/contract/types.rs:47]: I think in more recent ABI output, tuples have names. This would require `ethabi` support to parse this value.

**Comment by nlordell** [comment]
I found an old PR for adding support for decoding nested tuples:
https://github.com/rust-ethereum/ethabi/pull/186

File: examples/truffle/contracts/AbiTypes.sol
Comment (nlordell) [reply examples/truffle/contracts/AbiTypes.sol:90]: I'm not sure I understand why the `int`s don't work. If the value is out of range, its OK:
```rust
    let x = 0xffffffff_ffffffff_ffffffff_ffffffd6u128;
    dbg!(x as i8); // -42
```

And, AFAIK, Solidity ABI ensures these values are sign-extended (and even if they weren't I'm still not sure I understand the problem).

File: ethcontract/src/tokens.rs
Comment (nlordell) [reply ethcontract/src/tokens.rs]: I'd also be OK with moving `ethabi` to a `ethcontract-abi` and modifying it to better suit our needs. If I remember correctly, there are a few things that we end up having to work around because our needs don't fit very well with the library. Off the top of my head:
- Creating a lookup of [function name -> ABI fragment (this causes problems with overloaded function)](https://github.com/gnosis/ethcontract-rs/blob/8214590b791af0051c439918850247acdc098dd9/ethcontract/src/contract.rs#L318)
- for full v0.6 support we need to [pre-process the JSON](https://github.com/gnosis/ethcontract-rs/blob/8214590b791af0051c439918850247acdc098dd9/ethcontract-common/src/truffle.rs#L112)

File: examples/truffle/contracts/AbiTypes.sol
Comment (vkgnosis) [reply examples/truffle/contracts/AbiTypes.sol:90]: In the solidity abi an integer with less 256 bits is still encoded as a u256 or i256. For unsigned integers the value is 0 extended and for signed integers sign extended.

The rust-web3 code does not enforce this as it always takes the low bits without looking at the rest.
My new code does enforce this.

This surfaced a new bug in ethabi where unnamed event variables are not encoded properly https://github.com/rust-ethereum/ethabi/issues/206 .

Comment (nlordell) [reply examples/truffle/contracts/AbiTypes.sol:90]: Ah, this makes perfect sense. I guess the issue is that it was too permissive before and its probably better to range check the downcasting from the `U256` to `[ui]N`

File: ethcontract/src/tokens.rs
Comment (nlordell) [reply ethcontract/src/tokens.rs:69]: I agree, I think it should only accept `[u8; N]` for fixed bytes.

File: ethcontract/src/tokens.rs
Comment (nlordell) [inline ethcontract/src/tokens.rs:321]: We should re-write this for const-generics once it lands. [Its currently in beta](https://blog.rust-lang.org/2021/02/26/const-generics-mvp-beta).

File: ethcontract/src/tokens.rs
Comment (nlordell) [inline ethcontract/src/tokens.rs:34]: I like this decision to put the error in this module instead of the `errors` module. I wish I had started that way for the other `*Error` types I had :stuck_out_tongue:

File: ethcontract/src/tokens.rs
Comment (nlordell) [inline ethcontract/src/tokens.rs]: I wonder if, as you mentioned, we could just use `SingleTokenize` and then have a `TryInto<Vec<Token>>` for unwrapping `Token::Tuple` variants instead of having a separate trait. This would require all function calls to be called with tuple arguments, but I think think this is fine since we generate the code anyway.

I might be missing something though.

**Review by nlordell — COMMENTED** [review]
Code looks good to me.

It might be worth exploring whether the `MultiTokenize` can just be represented by a `SingleTokenize` that always returns a tuple, this way we have one less trait.

I think the distinction was important in `web3` prior to ABI v2 as function parameters are kind of like a tuple, but nested tuples weren't allowed, but since this is no longer the case, we might get away with a single trait.

**Comment by vkgnosis** [comment]
Added two new commits that simplify the code by removing MultiTokenize and Return. It turns out we can represent both cases as SingleTokenize tuples.

**Comment by vkgnosis** [comment]
The one remaining point is what to do about the ethabi fork. As Nick pointed out we can't use the git dependency because it would prevent ethcontract from being publishes to crates.io. I think a reasonable short term fix is if I publish a version of my fork so we can have it as a crates.io dependency. Long term we can still depends on the original ethabi crate.

File: ethcontract-generate/src/contract/methods.rs
Comment (nlordell) [inline ethcontract-generate/src/contract/methods.rs:148]: I'm very happy this isn't needed anymore :tada:

File: ethcontract/src/contract/method.rs
Comment (nlordell) [inline ethcontract/src/contract/method.rs]: Maybe this can be part of the `Tokenize` trait as a default method?

Although, on second thought, this might no longer be needed with #477, so maybe just leave it as it currently is.

**Review by nlordell — APPROVED** [review]
Amazing! I think it turned out really good!

**Comment by vkgnosis** [comment]
@nlordell FYI I had to add another small commit (the most recent one) here for a problem I discovered when using this branch in v2-services. It's a minor change and I am probably going to merge this PR tomorrow.

**Review by nlordell — APPROVED** [review]
Changes look fine to me.

**Pull request merged**
